[sql lab] allow users to save their queries

mistercrunch opened the pull request:
![img](http://g.recordit.co/pKtVtLyv9r.gif)
* new SavedQuery model & ModelView
* start moving sql_lab logic into `models/sql_lab.py` and `views/sql_lab.py` 
* introducing `notify` alerts using `react-alert`

mistercrunch committed:
Allow users to save their queries

Fixing tests .

ascott commented on superset/assets/javascripts/SqlLab/components/SaveQuery.jsx:
```diff
@@ -0,0 +1,139 @@
+      db_id: this.props.dbId,
+      schema: this.props.schema,
+      sql: this.props.sql,
+      csrf_token: $('input#csrf_token').val(),
```
this seems like an anti-pattern... why are we sending the csrf token here?

ascott replied:
i think a better pattern here might be to send this as a header on the xhr request.
http://flask-wtf.readthedocs.io/en/stable/csrf.html#javascript-requests

ascott commented on superset/models/sql_lab.py:
```diff
@@ -0,0 +1,159 @@
+
+    def to_dict(self):
+        return {
+            'changedOn': self.changed_on,
```
do we need both `changedOn` and `changed_on`?

mistercrunch replied:
Just moving the class from a module to another here as an effort to break down `models.py` that was humongous, we should get to the bottom of it in another PR though.

ascott commented on superset/assets/javascripts/SqlLab/actions.js:
```diff
@@ -326,8 +327,31 @@ export function popStoredQuery(urlId) {
+          sql: sq.sql,
+        };
+        dispatch(addQueryEditor(queryEditorProps));
+        notify.success('Query loaded in a new tab from link');
```
i don't think we need to add a success notification here since the user will see that this has happened in the ui. it's only on error that the notification is important here i think.

ascott commented on superset/assets/javascripts/SqlLab/components/SaveQuery.jsx:
```diff
@@ -0,0 +1,139 @@
+    this.onDescriptionChange = this.onDescriptionChange.bind(this);
+  }
+  onSave() {
+    const url = '/savedqueryviewapi/api/create';
```
we should put this request in `actions.js` https://github.com/ascott/superset/blob/master/superset/assets/javascripts/SqlLab/actions.js

ascott commented on superset/assets/javascripts/SqlLab/components/SaveQuery.jsx:
```diff
@@ -0,0 +1,139 @@
+          {this.renderPopover()}
+        </Overlay>
+        <Button bsSize="small" className="toggleSave" onClick={this.toggleSave}>
+          <i className="fa fa-save" /> Save
```
i think it would be more clear to users if we labeled this `Save Query`

ascott commented on superset/assets/javascripts/SqlLab/components/TabbedSqlEditors.jsx:
```diff
@@ -6,7 +6,7 @@ import * as Actions from '../actions';
 import CopyQueryTabUrl from './CopyQueryTabUrl';
 import { areArraysShallowEqual } from '../../reduxUtils';
-import { getParamFromQuery } from '../../../utils/common';
+import URI from 'urijs';
```
what is the advantage of `urijs` over using `getParamFromQuery`? it does look a little cleaner, but i'm thinking it might not be worth the addition depending on the package size.
how big of a lib is `urijs`?

ascott replied:
the sql lab entry js file 13.2mb 🙈 
![screenshot 2017-04-03 17 46 22](https://cloud.githubusercontent.com/assets/130878/24637381/8f04fcc4-1895-11e7-9f6a-cde4a3afa018.png)

mistercrunch replied:
I really need it for this PR though: https://github.com/airbnb/superset/pull/2537 

We should dig into what is taking those 13 mb. I'll try to dig into that, I bet urijs doesn't really weight in..

ascott commented on superset/assets/javascripts/SqlLab/components/SaveQuery.jsx:
```diff
@@ -0,0 +1,139 @@
+      type: 'POST',
+      url,
+      data,
+      success: () => notify.success('The query was saved'),
```
Suggested edit: `s/The/Your`

ascott commented on superset/assets/javascripts/SqlLab/components/SaveQuery.jsx:
```diff
@@ -0,0 +1,139 @@
+      url,
+      data,
+      success: () => notify.success('The query was saved'),
+      error: e => notify.error(`The query couldn't be saved. \n${e}`),
```
Suggested edit: s/The/Your

ascott commented:
this is looking good!

can we use less vertical padding in the alert like the default theme? 

![screenshot 2017-04-03 18 07 57](https://cloud.githubusercontent.com/assets/130878/24637798/b075ecbc-1898-11e7-8e45-d70eac24c016.png)
![screenshot 2017-04-03 18 07 44](https://cloud.githubusercontent.com/assets/130878/24637799/b0772d5c-1898-11e7-98e9-eeb182b30e8d.png)

ascott commented on superset/assets/javascripts/SqlLab/components/AlertsWrapper.jsx:
```diff
@@ -0,0 +1,18 @@
+        position="top right"
+        theme="dark"
+        time={5000}
+        transition="scale"
```
i'm partial to the fade rather than scale for notifications.

mistercrunch commented:
@ascott I addressed all comments except for the cosmetics on alerts as I'd rather let you alter the alerts to your liking [my css skills could only make things look worse...].

I'd like to stick with urijs as used it in a PR on top of this one. I'll look into what's causing that 14mb though.

ascott commented on superset/assets/javascripts/SqlLab/components/SaveQuery.jsx:
```diff
@@ -0,0 +1,131 @@
+              </small>
+              <FormControl
+                componentClass="textarea"
+                placeholder="textarea"
```
small nit: can we change this placeholder to 'Write a description for your query'

mistercrunch replied:
oops

ascott commented on superset/assets/javascripts/SqlLab/index.jsx:
```diff
@@ -14,6 +15,7 @@ import App from './components/App';
 
 
 require('./main.css');
+initJQueryCSRF();
```
nit: `initAjaxSetup` might be more clear here...

mistercrunch committed:
Adding placeholder for Query Description

ascott commented:
👍  lgtm

mistercrunch committed:
initJQueryCSRF -> initJQueryAjaxCSRF
